// @discord-player/extractor/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.version = exports.Lyrics = exports.Vimeo = exports.Reverbnation = exports.Facebook = exports.Attachment = exports.internal = void 0;
const Attachment_1 = require("./internal/Attachment");
const Facebook_1 = require("./internal/Facebook");
const Vimeo_1 = require("./internal/Vimeo");
exports.internal = { Attachment: Attachment_1.Attachment, Facebook: Facebook_1.Facebook, Vimeo: Vimeo_1.Vimeo };
exports.Attachment = __importStar(require("./Attachment"));
exports.Facebook = __importStar(require("./Facebook"));
exports.Reverbnation = __importStar(require("./Reverbnation"));
exports.Vimeo = __importStar(require("./Vimeo"));
exports.Lyrics = __importStar(require("./ext/Lyrics"));
exports.version = require('../package.json').version;
